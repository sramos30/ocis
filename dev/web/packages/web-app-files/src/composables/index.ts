export * from './fileListHeaderPosition'
export * from './pagination'
export * from './resourcesViewDefaults'
export * from './router'
export * from './sort'

declare module 'vue/types/vue' {
  interface Vue {
    [key: string]: any
  }
}
