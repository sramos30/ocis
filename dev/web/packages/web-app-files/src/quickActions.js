import { DateTime } from 'luxon'
import copyToClipboard from 'copy-to-clipboard'

export function createPublicLink(ctx) {
  // FIXME: Translate name
  const params = { name: ctx.$gettext('Quick action link'), permissions: 1 }
  const capabilities = ctx.store.state.user.capabilities
  const expirationDate = capabilities.files_sharing.public.expire_date

  if (expirationDate.enabled) {
    params.expireDate = DateTime.now()
      .plus({ days: parseInt(expirationDate.days, 10) })
      .endOf('day')
      .toISO()
  }

  if (ctx.storageId) {
    params.spaceRef = `${ctx.storageId}${ctx.item.path}`
  }

  return new Promise((resolve, reject) => {
    ctx.store
      .dispatch('Files/addLink', {
        path: ctx.item.path,
        client: ctx.client,
        params,
        storageId: ctx.storageId
      })
      .then((link) => {
        ctx.store.dispatch('Files/sidebar/openWithPanel', 'links-item').then(() => {
          copyToClipboard(link.url)
          ctx.store.dispatch('showMessage', {
            title: ctx.$gettext('Public link created'),
            desc: ctx.$gettext(
              'Public link was created successfully and copied into your clipboard.'
            )
          })
        })
        resolve()
      })
      .catch((e) => {
        reject(e)
      })
  })
}

export async function openNewCollaboratorsPanel(ctx) {
  await ctx.store.dispatch('Files/sidebar/openWithPanel', 'sharing-item')
}

export async function openSpaceMembersPanel(ctx) {
  await ctx.store.dispatch('Files/sidebar/openWithPanel', 'space-share-item')
}

export function canShare(item, store) {
  return (
    store.state.user.capabilities.files_sharing &&
    store.state.user.capabilities.files_sharing.api_enabled &&
    item.canShare()
  )
}

export default {
  collaborators: {
    id: 'collaborators',
    label: ($gettext) => $gettext('Add people'),
    icon: 'user-add',
    handler: openNewCollaboratorsPanel,
    displayed: canShare
  },
  publicLink: {
    id: 'public-link',
    label: ($gettext) => $gettext('Create and copy public link'),
    icon: 'link',
    handler: createPublicLink,
    displayed: canShare
  }
}
