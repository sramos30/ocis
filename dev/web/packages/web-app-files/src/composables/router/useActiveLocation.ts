import { ref, Ref, watch } from '@vue/composition-api'
import { useRoute, useRouter } from 'web-pkg/src/composables'
import { ActiveRouteDirectorFunc } from '../../router'

/**
 * watches the current route and re-evaluates the provided active location director
 * on each route name update.
 *
 * @param director
 * @param comparatives
 */
export const useActiveLocation = <T extends string>(
  director: ActiveRouteDirectorFunc<T>,
  ...comparatives: T[]
): Ref<boolean> => {
  const value = ref(false)
  const router = useRouter()
  const currentRoute = useRoute()
  watch(
    currentRoute,
    () => {
      value.value = director(router, ...comparatives)
    },
    { immediate: true }
  )
  return value
}
