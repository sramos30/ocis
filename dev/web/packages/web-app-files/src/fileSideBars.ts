import FileDetails from './components/SideBar/Details/FileDetails.vue'
import FileDetailsMultiple from './components/SideBar/Details/FileDetailsMultiple.vue'
import FileActions from './components/SideBar/Actions/FileActions.vue'
import FileVersions from './components/SideBar/Versions/FileVersions.vue'
import FileShares from './components/SideBar/Shares/FileShares.vue'
import FileLinks from './components/SideBar/Links/FileLinks.vue'
import NoSelection from './components/SideBar/NoSelection.vue'
import SpaceActions from './components/SideBar/Actions/SpaceActions.vue'
import SpaceDetails from './components/SideBar/Details/SpaceDetails.vue'
import SpaceShares from './components/SideBar/Shares/SpaceShares.vue'
import { isLocationSpacesActive, isLocationTrashActive, isLocationPublicActive } from './router'
import { spaceRoleEditor, spaceRoleManager } from './helpers/share'
import { Panel } from '../../web-pkg/src/components/sidebar'

import { Resource } from './helpers/resource'
import { User } from './helpers/user'
import Router from 'vue-router'

function $gettext(msg: string): string {
  return msg
}

const panelGenerators: (({
  rootFolder,
  highlightedFile,
  router,
  multipleSelection,
  user,
  capabilities
}: {
  rootFolder: boolean
  highlightedFile: Resource
  router: Router
  multipleSelection: boolean
  user: User
  capabilities: any
}) => Panel)[] = [
  // We don't have file details in the trashbin, yet.
  // Only allow `actions` panel on trashbin route for now.
  ({ rootFolder, highlightedFile }): Panel => ({
    app: 'no-selection-item',
    icon: 'questionnaire-line',
    title: $gettext('Details'),
    component: NoSelection,
    default: () => true,
    get enabled() {
      return rootFolder && highlightedFile?.type !== 'space'
    }
  }),
  ({ router, multipleSelection, rootFolder }) => ({
    app: 'details-item',
    icon: 'questionnaire-line',
    title: $gettext('Details'),
    component: FileDetails,
    default:
      !isLocationTrashActive(router, 'files-trash-personal') &&
      !isLocationTrashActive(router, 'files-trash-spaces-project'),
    get enabled() {
      return (
        !isLocationTrashActive(router, 'files-trash-personal') &&
        !isLocationTrashActive(router, 'files-trash-spaces-project') &&
        !multipleSelection &&
        !rootFolder
      )
    }
  }),
  ({ multipleSelection, rootFolder }) => ({
    app: 'details-multiple-item',
    icon: 'questionnaire-line',
    title: $gettext('Details'),
    component: FileDetailsMultiple,
    default: () => true,
    get enabled() {
      return multipleSelection && !rootFolder
    }
  }),
  ({ router, highlightedFile }) => ({
    app: 'details-space-item',
    icon: 'questionnaire-line',
    title: $gettext('Details'),
    component: SpaceDetails,
    default: isLocationSpacesActive(router, 'files-spaces-projects'),
    get enabled() {
      return highlightedFile?.type === 'space'
    }
  }),
  ({ router, multipleSelection, rootFolder }) => ({
    app: 'actions-item',
    icon: 'slideshow-3',
    title: $gettext('Actions'),
    component: FileActions,
    default:
      isLocationTrashActive(router, 'files-trash-personal') ||
      isLocationTrashActive(router, 'files-trash-spaces-project'),
    get enabled() {
      return !multipleSelection && !rootFolder
    }
  }),
  ({ highlightedFile, user }) => ({
    app: 'space-actions-item',
    icon: 'slideshow-3',
    title: $gettext('Actions'),
    component: SpaceActions,
    get enabled() {
      if (highlightedFile?.type !== 'space') {
        return false
      }
      return [
        ...highlightedFile.spaceRoles[spaceRoleManager.name],
        ...highlightedFile.spaceRoles[spaceRoleEditor.name]
      ].includes(user.uuid)
    }
  }),
  ({ capabilities, router, multipleSelection, rootFolder }) => ({
    app: 'sharing-item',
    icon: 'group',
    title: $gettext('People'),
    component: FileShares,
    get enabled() {
      if (multipleSelection || rootFolder) return false
      if (
        isLocationTrashActive(router, 'files-trash-personal') ||
        isLocationTrashActive(router, 'files-trash-spaces-project') ||
        isLocationPublicActive(router, 'files-public-files')
      ) {
        return false
      }

      if (capabilities.files_sharing) {
        return capabilities.files_sharing.api_enabled
      }
      return false
    }
  }),
  ({ highlightedFile }) => ({
    app: 'space-share-item',
    icon: 'group',
    title: $gettext('Members'),
    component: SpaceShares,
    get enabled() {
      return highlightedFile?.type === 'space'
    }
  }),
  ({ capabilities, router, multipleSelection, rootFolder, highlightedFile }) => ({
    app: 'links-item',
    icon: 'link',
    title: $gettext('Links'),
    component: FileLinks,
    get enabled() {
      if (multipleSelection || (rootFolder && highlightedFile?.type !== 'space')) return false
      if (
        isLocationTrashActive(router, 'files-trash-personal') ||
        isLocationTrashActive(router, 'files-trash-spaces-project') ||
        isLocationPublicActive(router, 'files-public-files')
      ) {
        return false
      }

      if (capabilities.files_sharing) {
        return capabilities.files_sharing.public.enabled
      }
      return false
    }
  }),
  ({ capabilities, highlightedFile, router, multipleSelection, rootFolder }) => ({
    app: 'versions-item',
    icon: 'git-branch',
    title: $gettext('Versions'),
    component: FileVersions,
    get enabled() {
      if (multipleSelection || rootFolder) return false
      if (
        isLocationTrashActive(router, 'files-trash-personal') ||
        isLocationTrashActive(router, 'files-trash-spaces-project') ||
        isLocationPublicActive(router, 'files-public-files')
      ) {
        return false
      }
      return !!capabilities.core && highlightedFile && highlightedFile.type !== 'folder'
    }
  })
]

export default panelGenerators
