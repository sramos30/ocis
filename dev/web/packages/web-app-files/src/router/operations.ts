import { RouteComponents } from './router'
import { Location, RouteConfig } from 'vue-router'
import { $gettext, createLocation, isLocationActiveDirector } from './utils'

type operationsTypes =
  | 'files-operations-location-picker'
  | 'files-operations-resolver-private-link'
  | 'files-operations-resolver-public-link'

export const createLocationOperations = (name: operationsTypes, location = {}): Location =>
  createLocation(name, location)

const locationLocationPicker = createLocationOperations('files-operations-location-picker')
const locationResolverPublicLink = createLocationOperations('files-operations-resolver-public-link')
const locationResolverPrivateLink = createLocationOperations(
  'files-operations-resolver-private-link'
)

export const isLocationOperationsActive = isLocationActiveDirector<operationsTypes>(
  locationLocationPicker,
  locationResolverPublicLink,
  locationResolverPrivateLink
)

export const buildRoutes = (components: RouteComponents): RouteConfig[] => [
  {
    name: locationLocationPicker.name,
    path: '/ops/location-picker/:context/:action/:item*',
    component: components.LocationPicker,
    meta: {
      verbose: true,
      auth: false,
      patchCleanPath: true
    }
  },
  {
    name: locationResolverPublicLink.name,
    path: '/ops/resolver/public-link/:token',
    component: components.PublicLink,
    meta: {
      auth: false,
      title: $gettext('Resolving public link')
    }
  },
  {
    name: locationResolverPrivateLink.name,
    path: '/ops/resolver/private-link/:fileId',
    component: components.PrivateLink,
    meta: { title: $gettext('Resolving private link') }
  }
]
