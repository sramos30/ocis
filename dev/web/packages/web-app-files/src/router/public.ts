import { RouteComponents } from './router'
import { Location, RouteConfig } from 'vue-router'
import { createLocation, isLocationActiveDirector, $gettext } from './utils'

type shareTypes = 'files-public-files' | 'files-public-drop'

export const createLocationPublic = (name: shareTypes, location = {}): Location =>
  createLocation(name, location)

export const locationPublicFiles = createLocationPublic('files-public-files')
export const locationPublicDrop = createLocationPublic('files-public-drop')

export const isLocationPublicActive = isLocationActiveDirector<shareTypes>(
  locationPublicFiles,
  locationPublicDrop
)

export const buildRoutes = (components: RouteComponents): RouteConfig[] => [
  {
    path: '/public',
    component: components.App,
    meta: {
      auth: false
    },
    children: [
      {
        name: locationPublicFiles.name,
        path: 'show/:item*',
        component: components.PublicFiles,
        meta: {
          auth: false,
          title: $gettext('Public files'),
          patchCleanPath: true
        }
      }
    ]
  },
  {
    name: locationPublicDrop.name,
    path: '/public/drop/:token',
    component: components.FilesDrop,
    meta: { auth: false, title: $gettext('Public file upload') }
  }
]
