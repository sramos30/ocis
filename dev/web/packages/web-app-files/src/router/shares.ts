import { RouteComponents } from './router'
import { Location, RouteConfig } from 'vue-router'
import { createLocation, isLocationActiveDirector, $gettext } from './utils'

type shareTypes = 'files-shares-with-me' | 'files-shares-with-others' | 'files-shares-via-link'

export const createLocationShares = (name: shareTypes, location = {}): Location =>
  createLocation(name, location)

export const locationSharesWithMe = createLocationShares('files-shares-with-me')
export const locationSharesWithOthers = createLocationShares('files-shares-with-others')
export const locationSharesViaLink = createLocationShares('files-shares-via-link')

export const isLocationSharesActive = isLocationActiveDirector<shareTypes>(
  locationSharesWithMe,
  locationSharesWithOthers,
  locationSharesViaLink
)

export const buildRoutes = (components: RouteComponents): RouteConfig[] => [
  {
    path: '/shares',
    component: components.App,
    redirect: locationSharesWithMe,
    children: [
      {
        name: locationSharesWithMe.name,
        path: 'with-me',
        component: components.SharedWithMe,
        meta: {
          title: $gettext('Files shared with me')
        }
      },
      {
        name: locationSharesWithOthers.name,
        path: 'with-others',
        component: components.SharedWithOthers,
        meta: {
          title: $gettext('Files shared with others')
        }
      },
      {
        name: locationSharesViaLink.name,
        path: 'via-link',
        component: components.SharedViaLink,
        meta: {
          title: $gettext('Files shared via link')
        }
      }
    ]
  }
]
