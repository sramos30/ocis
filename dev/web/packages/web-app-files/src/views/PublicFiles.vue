<template>
  <div>
    <app-bar
      :has-bulk-actions="true"
      :breadcrumbs="breadcrumbs"
      :breadcrumbs-context-actions-items="[currentFolder]"
    >
      <template #actions>
        <create-and-upload />
      </template>
    </app-bar>
    <app-loading-spinner v-if="loadResourcesTask.isRunning" />
    <template v-else>
      <progress-bar v-show="$_uploadProgressVisible" id="files-upload-progress" class="oc-p-s" />
      <not-found-message v-if="folderNotFound" class="files-not-found oc-height-1-1" />
      <no-content-message
        v-else-if="isEmpty"
        id="files-public-list-empty"
        class="files-empty"
        icon="folder"
      >
        <template #message>
          <span v-translate>There are no resources in this folder</span>
        </template>
        <template v-if="currentFolder.canCreate()" #callToAction>
          <span v-translate data-testid="public-files-call-to-action">
            Drag files and folders here or use the "New" or "Upload" buttons to add files
          </span>
        </template>
      </no-content-message>
      <resource-table
        v-else
        id="files-public-files-table"
        v-model="selectedResources"
        class="files-table"
        :class="{ 'files-table-squashed': !sidebarClosed }"
        :are-thumbnails-displayed="displayThumbnails"
        :resources="paginatedResources"
        :target-route="targetRoute"
        :header-position="fileListHeaderY"
        :sort-by="sortBy"
        :sort-dir="sortDir"
        @fileClick="$_fileActions_triggerDefaultAction"
        @rowMounted="rowMounted"
        @sort="handleSort"
      >
        <template #contextMenu="{ resource }">
          <context-actions v-if="isResourceInSelection(resource)" :items="selectedResources" />
        </template>
        <template #footer>
          <pagination :pages="paginationPages" :current-page="paginationPage" />
          <list-info
            v-if="paginatedResources.length > 0"
            class="oc-width-1-1 oc-my-s"
            :files="totalFilesCount.files"
            :folders="totalFilesCount.folders"
            :size="totalFilesSize"
          />
        </template>
      </resource-table>
    </template>
  </div>
</template>

<script lang="ts">
import { mapGetters, mapActions, mapMutations, mapState } from 'vuex'
import ResourceTable from '../components/FilesList/ResourceTable.vue'
import { useResourcesViewDefaults } from '../composables'

import MixinAccessibleBreadcrumb from '../mixins/accessibleBreadcrumb'
import MixinFileActions from '../mixins/fileActions'
import MixinMountSideBar from '../mixins/sidebar/mountSideBar'

import { VisibilityObserver } from 'web-pkg/src/observer'
import { ImageDimension, ImageType } from '../constants'
import debounce from 'lodash-es/debounce'
import { bus } from 'web-pkg/src/instance'

import AppLoadingSpinner from 'web-pkg/src/components/AppLoadingSpinner.vue'
import NoContentMessage from 'web-pkg/src/components/NoContentMessage.vue'
import AppBar from '../components/AppBar/AppBar.vue'
import ProgressBar from '../components/Upload/ProgressBar.vue'
import CreateAndUpload from '../components/AppBar/CreateAndUpload.vue'
import NotFoundMessage from '../components/FilesList/NotFoundMessage.vue'
import ListInfo from '../components/FilesList/ListInfo.vue'
import Pagination from '../components/FilesList/Pagination.vue'
import ContextActions from '../components/FilesList/ContextActions.vue'
import { createLocationOperations } from '../router'
import { breadcrumbsFromPath, concatBreadcrumbs } from '../helpers/breadcrumbs'
import { defineComponent } from '@vue/composition-api'
import { Resource } from '../helpers/resource'

const visibilityObserver = new VisibilityObserver()

export default defineComponent({
  components: {
    AppBar,
    ProgressBar,
    CreateAndUpload,
    ResourceTable,
    ListInfo,
    Pagination,
    AppLoadingSpinner,
    NoContentMessage,
    NotFoundMessage,
    ContextActions
  },

  mixins: [MixinAccessibleBreadcrumb, MixinFileActions, MixinMountSideBar],

  setup() {
    return {
      ...useResourcesViewDefaults<Resource, any, any[]>()
    }
  },

  computed: {
    ...mapGetters('Files', [
      'publicLinkPassword',
      'currentFolder',
      'highlightedFile',
      'inProgress',
      'totalFilesCount',
      'totalFilesSize'
    ]),
    ...mapGetters(['configuration']),
    ...mapState('Files/sidebar', { sidebarClosed: 'closed' }),

    $_uploadProgressVisible() {
      return this.inProgress.length > 0
    },

    breadcrumbs() {
      const breadcrumbs = breadcrumbsFromPath(this.$route.path, this.$route.params.item)
      const rootRoute = breadcrumbs.shift()

      return concatBreadcrumbs(
        { text: this.$gettext('Public link'), to: rootRoute.to },
        ...breadcrumbs
      )
    },

    isEmpty() {
      return this.paginatedResources.length < 1
    },

    folderNotFound() {
      return this.currentFolder === null
    },

    targetRoute() {
      return { name: this.$route.name }
    },

    displayThumbnails() {
      return !this.configuration.options.disablePreviews
    }
  },

  watch: {
    $route: {
      handler: function (to, from) {
        const sameRoute = to.name === from?.name
        const sameItem = to.params?.item === from?.params?.item
        if (!sameRoute || !sameItem) {
          this.loadResourcesTask.perform(this, sameRoute)
        }
      },
      immediate: true
    }
  },
  beforeDestroy() {
    visibilityObserver.disconnect()
  },

  created() {
    const loadResourcesEventToken = bus.subscribe('app.files.list.load', (path) => {
      this.loadResourcesTask.perform(this, this.$route.params.item === path, path)
    })

    this.$on('beforeDestroy', () => bus.unsubscribe('app.files.list.load', loadResourcesEventToken))
  },

  methods: {
    ...mapActions('Files', ['loadPreview']),
    ...mapMutations('Files', ['SET_CURRENT_FOLDER', 'LOAD_FILES', 'CLEAR_CURRENT_FILES_LIST']),

    rowMounted(resource, component) {
      if (!this.displayThumbnails) {
        return
      }

      const debounced = debounce(({ unobserve }) => {
        unobserve()
        this.loadPreview({
          resource,
          isPublic: true,
          dimensions: ImageDimension.Thumbnail,
          type: ImageType.Thumbnail
        })
      }, 250)

      visibilityObserver.observe(component.$el, { onEnter: debounced, onExit: debounced.cancel })
    },

    redirectToResolvePage() {
      this.$router.push(
        createLocationOperations('files-operations-resolver-public-link', {
          params: { token: this.$route.params.item }
        })
      )
    }
  }
})
</script>
