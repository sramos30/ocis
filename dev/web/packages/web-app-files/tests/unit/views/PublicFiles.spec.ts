import { mount, RouterLinkStub } from '@vue/test-utils'
import { accentuatesTableRowTest } from './views.shared'
import PublicFiles from '@files/src/views/PublicFiles.vue'
import { localVue, getStore, createFile } from './views.setup'
import { createLocationPublic } from '../../../src/router'

const createFolder = ({ name = '1234', canCreate = false } = {}) => ({
  name: name,
  canCreate: () => canCreate
})

const stubs = {
  'app-bar': true,
  'app-loading-spinner': true,
  'create-and-upload': true,
  'not-found-message': true,
  'no-content-message': true,
  'progress-bar': true,
  'resource-table': true,
  'context-actions': true,
  pagination: true,
  'list-info': true,
  'router-link': RouterLinkStub
}

const router = {
  push: jest.fn(),
  afterEach: jest.fn(),
  currentRoute: {
    ...createLocationPublic('files-public-files'),
    query: {}
  },
  resolve: (r) => {
    return { href: r.name }
  }
}

const resourceList = [createFile({ id: 1234 }), createFile({ id: 1235 }), createFile({ id: 1236 })]

const stubSelectors = {
  appLoadingSpinner: 'app-loading-spinner-stub',
  notFoundMessage: 'not-found-message-stub',
  noContentMessage: 'no-content-message-stub',
  resourceTable: 'resource-table-stub',
  contextActions: 'context-actions-stub',
  pagination: 'pagination-stub',
  listInfo: 'list-info-stub'
}

describe('PublicFiles view', () => {
  describe('accentuate new files and folders', () => {
    // eslint-disable-next-line jest/expect-expect
    it('accentuates table row for new files, folders and versions [Files/UPSERT_RESOURCE]', async () => {
      await accentuatesTableRowTest(PublicFiles)
    })
  })
  it('should show the app-loading-spinner component when the view is still loading', () => {
    const wrapper = getMountedWrapper({ loading: true })

    expect(wrapper.find(stubSelectors.appLoadingSpinner).exists()).toBeTruthy()
    expect(wrapper.find(stubSelectors.noContentMessage).exists()).toBeFalsy()
    expect(wrapper.find(stubSelectors.resourceTable).exists()).toBeFalsy()
  })
  describe('when the view is not loading anymore', () => {
    it('should not show the app-loading-spinner component', () => {
      const wrapper = getMountedWrapper({ loading: false })

      expect(wrapper.find(stubSelectors.appLoadingSpinner).exists()).toBeFalsy()
    })

    describe('not found message component', () => {
      it('should be visible if "folderNotFound" is set as "true"', () => {
        const wrapper = getMountedWrapper({ loading: false })
        expect(wrapper.find(stubSelectors.notFoundMessage).exists()).toBeTruthy()
      })
      it('should not be visible if "folderNotFound" is set as "false"', () => {
        const wrapper = getMountedWrapper({
          loading: false,
          currentFolder: createFolder()
        })
        expect(wrapper.find(stubSelectors.notFoundMessage).exists()).toBeFalsy()
      })
    })

    describe('no content message component', () => {
      it('should be visible if the length of the paginated resources is zero', () => {
        const wrapper = getMountedWrapper({
          paginatedResources: [],
          currentFolder: {
            canCreate: jest.fn()
          }
        })
        expect(wrapper.find(stubSelectors.noContentMessage).exists()).toBeTruthy()
      })
      it.each([
        {
          canCreate: true,
          expectedToExist: true
        },
        {
          canCreate: false,
          expectedToExist: false
        }
      ])(
        'should show/hide the "call to action" information according to the currentFolder "canCreate" permission',
        (cases) => {
          stubs['no-content-message'] = false
          const wrapper = getMountedWrapper({
            paginatedResources: [],
            currentFolder: createFolder({ canCreate: cases.canCreate })
          })

          const noContentMessage = wrapper.find('#files-public-list-empty')
          const callToAction = noContentMessage.find('[data-testid="public-files-call-to-action"]')
          expect(callToAction.exists()).toBe(cases.expectedToExist)
          stubs['no-content-message'] = true
        }
      )
    })
    describe('when length of the paginated resources is greater than zero', () => {
      const selectedResourceList = [resourceList[0], resourceList[1]]
      const notSelectedResourceList = [resourceList[2]]

      it('should load the resource table with the correct props', () => {
        const wrapper = getMountedWrapper({
          paginatedResources: resourceList,
          currentFolder: createFolder()
        })
        const resourceTable = wrapper.find(stubSelectors.resourceTable)
        expect(resourceTable).toMatchSnapshot()
      })
      describe('context menu', () => {
        let wrapper
        beforeEach(() => {
          stubs['resource-table'] = false
          wrapper = getMountedWrapper({
            paginatedResources: resourceList,
            selectedFiles: selectedResourceList,
            currentFolder: createFolder()
          })
        })
        afterEach(() => {
          stubs['resource-table'] = true
        })
        it('should show the context actions for every selected resource', () => {
          selectedResourceList.forEach((selectedResource) => {
            const resourceRow = wrapper.find(`[data-item-id="${selectedResource.id}"]`)
            const contextActions = resourceRow.find(stubSelectors.contextActions)
            expect(contextActions.exists()).toBeTruthy()
            expect(contextActions.props()).toMatchObject({
              items: selectedResourceList
            })
          })
        })
        it('should not show the context actions for the resources that are not selected', () => {
          notSelectedResourceList.forEach((notSelectedResource) => {
            const resourceRow = wrapper.find(`[data-item-id="${notSelectedResource.id}"]`)
            const contextActions = resourceRow.find(stubSelectors.contextActions)
            expect(contextActions.exists()).toBeFalsy()
          })
        })
      })
      describe('pagination component inside the resource table', () => {
        it('should be visible if pagination pages is greater than zero', () => {
          stubs['resource-table'] = false
          const wrapper = getMountedWrapper({
            paginatedResources: resourceList,
            paginationPages: 22,
            paginationPage: 2,
            currentFolder: createFolder()
          })

          const paginationItem = wrapper.find(stubSelectors.pagination)

          expect(paginationItem.exists()).toBeTruthy()
          expect(paginationItem.props()).toMatchObject({
            pages: 22,
            currentPage: 2
          })
          stubs['resource-table'] = true
        })
      })
      describe('list info component inside the resource table', () => {
        it('should be visible if paginated resources list is empty', () => {
          stubs['resource-table'] = false
          const wrapper = getMountedWrapper({
            paginatedResources: resourceList,
            totalFilesCount: { files: 2, folders: 3 },
            currentFolder: createFolder()
          })

          const listInfoItem = wrapper.find(stubSelectors.listInfo)

          expect(listInfoItem.exists()).toBeTruthy()
          expect(listInfoItem.props()).toMatchObject({
            files: 2,
            folders: 3,
            size: null
          })
          stubs['resource-table'] = true
        })
      })
    })
  })

  function mountOptions({ store, loading, paginatedResources, paginationPages, paginationPage }) {
    return {
      localVue,
      store: store,
      stubs,
      mocks: {
        $route: router.currentRoute,
        $router: router
      },
      computed: {
        breadcrumbs: () => []
      },
      setup: () => ({
        loadResourcesTask: {
          isRunning: loading,
          perform: jest.fn()
        },
        paginatedResources: paginatedResources,
        paginationPages: paginationPages,
        paginationPage: paginationPage
      })
    }
  }

  function getMountedWrapper({
    currentFolder = null,
    selectedFiles = [],
    loading = false,
    paginatedResources = [],
    paginationPages = 1,
    paginationPage = 1,
    totalFilesCount = { files: 12, folders: 21 }
  } = {}) {
    const component = { ...PublicFiles, created: jest.fn() }
    const store = getStore({
      totalFilesCount: totalFilesCount,
      selectedFiles,
      currentFolder
    })
    return mount(
      component,
      mountOptions({
        store,
        loading,
        paginatedResources,
        paginationPages,
        paginationPage
      })
    )
  }
})
