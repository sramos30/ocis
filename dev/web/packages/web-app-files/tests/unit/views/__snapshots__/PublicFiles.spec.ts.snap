// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PublicFiles view when the view is not loading anymore when length of the paginated resources is greater than zero should load the resource table with the correct props 1`] = `<resource-table-stub resources="[object Object],[object Object],[object Object]" resourcedomselector="[Function]" selection="" hasactions="true" targetroute="[object Object]" areresourcesclickable="true" headerposition="0" isselectable="true" paddingx="small" hover="true" sortdir="desc" id="files-public-files-table" class="files-table files-table-squashed"></resource-table-stub>`;
